stdlib/asm/collections/smt.masm

```diff
@@ -0,0 +1,249 @@
+
+    # check the opening value
+    # cycles: 20
+    movupw.2 movup.12 push.32 mtree_get movupw.2
+    # => [V', N' R, V, ...]
+
+    # assert the opening matches the computed node value and return
+    # cycles: 12
+    assert_eqw swapw
+    # => [V, R, ...]
+end
+
+#! Get the leaf value for depth 48
+#!
+#! Input:  [V, K, R, ...]
+#! Output: [V, R, ...]
+#
+#! Cycles: 106
+proc.get_48
+    # compute the remaining path and index `i`
+    # cycles: 18
+    swapw exec.split_48 swap movdn.12
+    # => [K', V, R, i, ...]
+
+    # prepare the permutation argument for hash in domain
+    # cycles: 16
+    swapw dupw movdnw.2 push.0.48.0.0 movdnw.2
+    # => [V, K', D, V, R, i, ...]
+
+    # execute the permutation and preserve the digest
+    # cycles: 10
+    hperm dropw swapw dropw
+    # => [N, V, R, i, ...]
+
+    # set the flag if the value is zero
+    # cycles: 19
+    dup.7 eq.0 dup.7 eq.0 and dup.6 eq.0 and dup.6 eq.0 and
+    # => [V == 0, N, V, R, i, ...]
+
+    # conditionally select empty constant or node value
+    # cycles: 11
+    push.10650694022550988030.5634734408638476525
+    push.9233115969432897632.1437907447409278328
+    movup.4 cdropw
+    # => [N', V, R, i, ...]
+
+    # check the opening value
+    # cycles: 20
+    movupw.2 movup.12 push.48 mtree_get movupw.2
+    # => [V', N' R, V, ...]
+
+    # assert the opening matches the computed node value and return
+    # cycles: 12
+    assert_eqw swapw
+    # => [V, R, ...]
+end
+
+#! Get the leaf value for depth 64
+#!
+#! Input:  [V, K, R, ...]
+#! Output: [V, R, ...]
+#
+#! Cycles: 94
+proc.get_64
+    # compute the remaining path and index `i`
+    # cycles: 6
+    swapw movdn.11 push.0
+    # => [K', V, R, i, ...]
+
+    # prepare the permutation argument for hash in domain
+    # cycles: 16
+    swapw dupw movdnw.2 push.0.64.0.0 movdnw.2
+    # => [V, K', D, V, R, i, ...]
+
+    # execute the permutation and preserve the digest
+    # cycles: 10
+    hperm dropw swapw dropw
+    # => [N, V, R, i, ...]
+
+    # set the flag if the value is zero
+    # cycles: 19
+    dup.7 eq.0 dup.7 eq.0 and dup.6 eq.0 and dup.6 eq.0 and
+    # => [V == 0, N, V, R, i, ...]
+
+    # conditionally select empty constant or node value
+    # cycles: 11
+    padw movup.4 cdropw
+    # => [N', V, R, i, ...]
+
+    # check the opening value
+    # cycles: 20
+    movupw.2 movup.12 push.64 mtree_get movupw.2
+    # => [V', N' R, V, ...]
+
+    # assert the opening matches the computed node value and return
+    # cycles: 12
+    assert_eqw swapw
+    # => [V, R, ...]
+end
+
+#! Returns the value stored under the specified key in a Sparse Merkle tree with the specified root.
+#!
+#! If the value for a given key has not been set, the returned `V` will consist of all zeroes.
+#!
+#! The consumed cycles will depend on the tier depth, and will respect the following table:
+#!
+#! depth 16: 101 cycles
+#! depth 32:  94 cycles
+#! depth 48: 106 cycles
+#! depth 64:  94 cycles
```

Comment:
I'd move this to the bottom and replace `Cycles: 106` with it.

Also, we should add 4 to the cycle count of each procedure because `if/else` statements themselves consume 2 cycles each.

Reply:
Fixed